File: backend/app/api/api_v1/endpoints/sessions.py
"""Sessions API endpoints."""
from typing import List, Optional
from datetime import datetime
from fastapi import APIRouter, Query, HTTPException
from motor.motor_asyncio import AsyncIOMotorDatabase
from bson import ObjectId

from app.api.dependencies import CommonDeps
from app.schemas.session import Session, SessionDetail, SessionWithMessages
from app.schemas.common import PaginatedResponse
from app.services.session import SessionService
from app.core.exceptions import NotFoundError

router = APIRouter()


@router.get("/", response_model=PaginatedResponse[Session])
async def list_sessions(
    db: CommonDeps,
    project_id: Optional[str] = Query(None, description="Filter by project ID"),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    search: Optional[str] = Query(None, description="Search in session summaries"),
    start_date: Optional[datetime] = Query(None, description="Filter by start date"),
    end_date: Optional[datetime] = Query(None, description="Filter by end date"),
    sort_by: str = Query("started_at", regex="^(started_at|ended_at|message_count|total_cost)$"),
    sort_order: str = Query("desc", regex="^(asc|desc)$")
) -> PaginatedResponse[Session]:
    """List sessions with pagination and filtering."""
    service = SessionService(db)
    
    # Build filter
    filter_dict = {}
    if project_id:
        if not ObjectId.is_valid(project_id):
            raise HTTPException(status_code=400, detail="Invalid project ID")
        filter_dict["projectId"] = ObjectId(project_id)
    
    if search:
        filter_dict["$text"] = {"$search": search}
    
    if start_date:
        filter_dict["startedAt"] = {"$gte": start_date}
    
    if end_date:
        if "startedAt" in filter_dict:
            filter_dict["startedAt"]["$lte"] = end_date
        else:
            filter_dict["startedAt"] = {"$lte": end_date}
    
    # Get sessions
    sessions, total = await service.list_sessions(
        filter_dict=filter_dict,
        skip=skip,
        limit=limit,
        sort_by=sort_by,
        sort_order=sort_order
    )
    
    return PaginatedResponse(
        items=sessions,
        total=total,
        skip=skip,
        limit=limit,
        has_more=skip + limit < total
    )


@router.get("/{session_id}", response_model=SessionDetail)
async def get_session(
    session_id: str,
    db: CommonDeps,
    include_messages: bool = Query(False, description="Include first 10 messages")
) -> SessionDetail:
    """Get a specific session by ID."""
    service = SessionService(db)
    session = await service.get_session(session_id, include_messages=include_messages)
    
    if not session:
        raise NotFoundError("Session", session_id)
    
    return session


@router.get("/{session_id}/messages", response_model=SessionWithMessages)
async def get_session_messages(
    session_id: str,
    db: CommonDeps,
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=200)
) -> SessionWithMessages:
    """Get all messages for a session.
    
    Returns messages in chronological order with pagination.
    """
    service = SessionService(db)
    
    # Get session
    session = await service.get_session(session_id)
    if not session:
        raise NotFoundError("Session", session_id)
    
    # Get messages
    messages = await service.get_session_messages(
        session_id,
        skip=skip,
        limit=limit
    )
    
    return SessionWithMessages(
        session=session,
        messages=messages,
        skip=skip,
        limit=limit
    )


@router.get("/{session_id}/thread/{message_uuid}")
async def get_message_thread(
    session_id: str,
    message_uuid: str,
    db: CommonDeps,
    depth: int = Query(10, ge=1, le=100, description="Maximum thread depth")
) -> dict:
    """Get the conversation thread for a specific message.
    
    Returns the message and its parent/child messages up to the specified depth.
    """
    service = SessionService(db)
    thread = await service.get_message_thread(session_id, message_uuid, depth)
    
    if not thread:
        raise NotFoundError("Message thread", f"{session_id}/{message_uuid}")
    
    return thread


@router.post("/{session_id}/generate-summary")
async def generate_session_summary(
    session_id: str,
    db: CommonDeps
) -> dict:
    """Generate or regenerate a summary for a session.
    
    Uses the first and last few messages to create a concise summary.
    """
    service = SessionService(db)
    summary = await service.generate_summary(session_id)
    
    if not summary:
        raise NotFoundError("Session", session_id)
    
    return {"session_id": session_id, "summary": summary}

File: frontend/src/api/sessions.ts
import { apiClient } from './client';
import { PaginatedResponse, Session, Message } from './types';

export interface SessionsParams {
  projectId?: string;
  skip?: number;
  limit?: number;
  search?: string;
  startDate?: string;
  endDate?: string;
  sortBy?: 'started_at' | 'ended_at' | 'message_count' | 'total_cost';
  sortOrder?: 'asc' | 'desc';
}

export interface SessionDetail extends Session {
  modelsUsed: string[];
  firstMessage?: string;
  lastMessage?: string;
  messages?: Message[];
}

export interface SessionWithMessages {
  session: SessionDetail;
  messages: Message[];
  skip: number;
  limit: number;
}

export interface MessageThread {
  message: Message;
  parents: Message[];
  children: Message[];
  depth: number;
}

export const sessionsApi = {
  async listSessions(
    params: SessionsParams = {}
  ): Promise<PaginatedResponse<Session>> {
    const queryParams = new URLSearchParams();

    if (params.projectId) queryParams.append('project_id', params.projectId);
    if (params.skip !== undefined)
      queryParams.append('skip', params.skip.toString());
    if (params.limit !== undefined)
      queryParams.append('limit', params.limit.toString());
    if (params.search) queryParams.append('search', params.search);
    if (params.startDate) queryParams.append('start_date', params.startDate);
    if (params.endDate) queryParams.append('end_date', params.endDate);
    if (params.sortBy) queryParams.append('sort_by', params.sortBy);
    if (params.sortOrder) queryParams.append('sort_order', params.sortOrder);

    return apiClient.get<PaginatedResponse<Session>>(
      `/sessions?${queryParams.toString()}`
    );
  },

  async getSession(
    sessionId: string,
    includeMessages = false
  ): Promise<SessionDetail> {
    const params = includeMessages ? '?include_messages=true' : '';
    return apiClient.get<SessionDetail>(`/sessions/${sessionId}${params}`);
  },

  async getSessionMessages(
    sessionId: string,
    skip = 0,
    limit = 100
  ): Promise<SessionWithMessages> {
    return apiClient.get<SessionWithMessages>(
      `/sessions/${sessionId}/messages?skip=${skip}&limit=${limit}`
    );
  },

  async getMessageThread(
    sessionId: string,
    messageUuid: string,
    depth = 10
  ): Promise<MessageThread> {
    return apiClient.get<MessageThread>(
      `/sessions/${sessionId}/thread/${messageUuid}?depth=${depth}`
    );
  },

  async generateSessionSummary(
    sessionId: string
  ): Promise<{ summary: string }> {
    const response = await apiClient.post<{
      session_id: string;
      summary: string;
    }>(`/sessions/${sessionId}/generate-summary`);
    return { summary: response.summary };
  },

  async updateMessageCost(messageId: string, costUsd: number) {
    const response = await apiClient.patch(
      `/messages/${messageId}/cost?cost_usd=${costUsd}`
    );
    return response;
  },

  async batchUpdateMessageCosts(costUpdates: Record<string, number>) {
    const response = await apiClient.post(
      '/messages/batch-update-costs',
      costUpdates
    );
    return response;
  },

  async forkSession(
    sessionId: string,
    messageId: string,
    description?: string
  ): Promise<{
    original_session_id: string;
    forked_session_id: string;
    forked_session_mongo_id: string;
    fork_point_message_id: string;
    description?: string;
    message_count: number;
  }> {
    const queryParams = new URLSearchParams();
    queryParams.append('message_id', messageId);
    if (description) {
      queryParams.append('description', description);
    }

    const response = await apiClient.post<{
      original_session_id: string;
      forked_session_id: string;
      forked_session_mongo_id: string;
      fork_point_message_id: string;
      description?: string;
      message_count: number;
    }>(`/sessions/${sessionId}/fork?${queryParams.toString()}`);
    return response;
  },
};
